aten/src/ATen/native/cuda/jit_utils.h

```diff
@@ -60,6 +60,22 @@ template <> inline std::string typeName<ctype>(){ \
     return std::string(#ctype);    \
 }
 
-AT_FORALL_SCALAR_TYPES_WITH_COMPLEX_AND_QINTS(TYPE_NAME_FN)
+AT_FORALL_SCALAR_TYPES(TYPE_NAME_FN)
+#undef TYPE_NAME_FN
+// JIT uses std::complex directly, because nvRTC compile programs
+// with -default-device, so there is no such issue like:
+//   "std::sin(complex) is __host__ only"
+template <> inline std::string typeName<c10::complex<float>>(){
+    return "std::complex<float>";
+}
+template <> inline std::string typeName<c10::complex<double>>(){
+    return "std::complex<double>";
+}
+template <> inline std::string typeName<at::Half>(){
```

Comment:
Why are at::Half and at::BFloat16 added here?

Reply:
(@ngimel wonders if they were left out on purpose, but we don't recall)

Reply:
Because I've changed the `AT_FORALL_SCALAR_TYPES_WITH_COMPLEX_AND_QINTS` into `AT_FORALL_SCALAR_TYPES` in previous lines, which does not include half and bfloat16. I can not use `AT_FORALL_SCALAR_TYPES_AND2(Half, BFloat16` either, because this will generate string `::c10::impl::ScalarTypeToCPPType<::c10::ScalarType::Half>::t` for Half, which is not what I want.